File: C++/deepest-leaves-sum.cpp
#include <vector>


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        // Time:  O(n)
        // Space: O(n)
        std::vector<TreeNode*> prev, current = {root};
        while (!current.empty()) {
            prev = std::move(current);
            for (const TreeNode* node: prev) {
                if (node) {
                    if (node->left) {
                        current.emplace_back(node->left);
                    }
                    if (node->right) {
                        current.emplace_back(node->right);
                    }
                }
            }
        }
        int result = 0;
        for (const TreeNode* node: prev) {
            result += node->val;
        }
        return result;
    }
};

File: Python/deepest-leaves-sum.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        # Time: O(n)
        # Space: O(n)
        prev, current = None, [root]
        while current:
            prev = current
            current = [child for node in current for child in (node.left, node.right) if child]
        return sum(node.val for node in prev)
